app/Jobs/Notifications/BroadcastNotificationBase.php

```diff
@@ -80,39 +81,51 @@ private static function applyDeliverySettings(array $userIds)
         static $defaults = ['mail' => true, 'push' => true];
 
         if (static::NOTIFICATION_OPTION_NAME !== null) {
-            // FIXME: filtering all the ids could get quite large?
-            $notificationOptions = UserNotificationOption
-                ::whereIn('user_id', $userIds)
-                ->where(['name' => static::NOTIFICATION_OPTION_NAME])
-                ->whereNotNull('details')
-                ->get()
-                ->keyBy('user_id');
-        } else {
-            $notificationOptions = [];
+            $notificationOptionsQuery = UserNotificationOption
+                ::where(['name' => static::NOTIFICATION_OPTION_NAME])
+                ->whereNotNull('details');
         }
 
         $deliverySettings = [];
-        foreach ($userIds as $userId) {
-            $details = $notificationOptions[$userId]->details ?? $defaults;
-            $delivery = 0;
-            foreach (UserNotification::DELIVERY_OFFSETS as $type => $_offset) {
-                if ($details[$type] ?? $defaults[$type]) {
-                    $delivery |= UserNotification::deliveryMask($type);
-                }
+
+        foreach (array_chunk($userIds, 10000) as $chunkedUserIds) {
+            if (isset($notificationOptionsQuery)) {
+                $notificationOptions = (clone $notificationOptionsQuery)
+                    ->whereIntegerInRaw('user_id', $chunkedUserIds)
+                    ->get()
+                    ->keyBy('user_id');
+            } else {
+                $notificationOptions = [];
             }
 
-            $deliverySettings[$userId] = $delivery;
+            foreach ($chunkedUserIds as $userId) {
+                $details = $notificationOptions[$userId]->details ?? $defaults;
+                $delivery = 0;
+                foreach (UserNotification::DELIVERY_OFFSETS as $type => $_offset) {
+                    if ($details[$type] ?? $defaults[$type]) {
+                        $delivery |= UserNotification::deliveryMask($type);
+                    }
+                }
+
+                $deliverySettings[$userId] = $delivery;
+            }
         }
 
         return $deliverySettings;
     }
 
     private static function excludeBotUserIds(array $userIds)
     {
-        // smaller return size from database compared to "group_id <> bot"
+        $botGroupId = app('groups')->byIdentifier('bot')->getKey();
+
+        $allBotUserIds = UserGroup
+            ::where('group_id', $botGroupId)
+            ->select('user_id');
+
+        // only consider users with bot as their primary group
         $botUserIds = User
-            ::whereIn('user_id', $userIds)
-            ->where('group_id', app('groups')->byIdentifier('bot')->getKey())
+            ::where('group_id', $botGroupId)
+            ->whereIn('user_id', $allBotUserIds)
```

Comment:
This part seems unnecessary - if going by the primary group, the `User::isBot` check is just `$this->group_id === app('groups')->byIdentifier('bot')->getKey();`
The old query was just limiting the result set to the list of users given instead of selecting all bot users.

I wonder if the list of bot users is something worth caching? 🤔

Reply:
yeah there's no index on `group_id` on users table. The query should be simple enough (primary key lookup on two tables) and the amount of data is minimal so it should be fine doing this query (or at least not much worse compared to the current one) :shrug: